infrastructure/src/main/java/org/corfudb/infrastructure/logreplication/FsmTaskManager.java

```diff
@@ -0,0 +1,275 @@
+
+            }
+        } catch (IllegalTransitionException illegalState) {
+            log.error("Illegal log replication event {} when in state {}", event.getType(), currState.getType());
+        }
+
+        synchronized(event.getRuntimeFsm()) {
+            sessionToRuntimeEventIdMap.get(session).remove(0);
+            event.getRuntimeFsm().notifyAll();
+        }
+    }
+
+    private boolean canProcessEventForSession(LogReplicationSession session, UUID currEventId) {
+            return (sessionToReplicationEventIdMap.get(session) != null &&
+                        !sessionToReplicationEventIdMap.get(session).isEmpty() &&
+                        sessionToReplicationEventIdMap.get(session).get(0).equals(currEventId)) ||
+                    (sessionToDelayedReplicationEventIdMap.get(session) != null &&
+                            !sessionToDelayedReplicationEventIdMap.get(session).isEmpty() &&
+                            sessionToDelayedReplicationEventIdMap.get(session).get(0).equals(currEventId));
+    }
+
+    private void processReplicationTask(LogReplicationEvent currEvent) {
+        LogReplicationSession session = currEvent.getReplicationFsm().getSession();
+        // for a given session, the fsm events should be processed in the order they are queued. This snippets ensures
+        // that in the event of 2 threads contending for the monitor, only the task submitted first would be processed.
+        synchronized(currEvent.getReplicationFsm()) {
+            while (true) {
+                if (canProcessEventForSession(session, currEvent.getEventId())) {
+                    break;
+                } else {
+                    try {
+                        currEvent.getReplicationFsm().wait();
+                    } catch (InterruptedException e) {
+                        log.error("Wait for session {} was interrupted {}", session, e.getMessage());
+                    }
+                }
+            }
+
+            LogReplicationState currState = currEvent.getReplicationFsm().getState();
+            if (currState.getType() == LogReplicationStateType.ERROR) {
+                log.info("Log Replication State Machine has been stopped. No more events will be processed.");
+                return;
+            }
+
+            // TODO (Anny): consider strategy for continuously failing snapshot sync (never ending cancellation)
+
+            try {
+                LogReplicationState newState = currState.processEvent(currEvent);
+                log.trace("Transition from {} to {}", currState, newState);
+
+                currEvent.getReplicationFsm().transition(currState, newState);
+                currEvent.getReplicationFsm().setState(newState);
+                currEvent.getReplicationFsm().getNumTransitions().setValue(currEvent.getReplicationFsm().getNumTransitions().getValue() + 1);
+
+            } catch (org.corfudb.infrastructure.logreplication.replication.fsm.IllegalTransitionException illegalState) {
```

Comment:
We have 2 IllegalTransitionException..one under replication and one under runtime. 
Since both the exceptions are used in this class, one of them will need to have the full path...unless we consolidate the exception